sosw/components/sigv4.py

```diff
@@ -0,0 +1,263 @@
+    of this software and associated documentation files (the "Software"), to deal
+    in the Software without restriction, including without limitation the rights
+    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+    copies of the Software, and to permit persons to whom the Software is
+    furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included in all
+    copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+"""
+
+__all__ = ['AWSSigV4RequestGenerator']
+__author__ = "Ivan Sushkov"
+
+import os
+import hmac
+import hashlib
+import logging
+import datetime
+
+from requests.auth import AuthBase
+from urllib.parse import quote, urlparse
+
+
+logger = logging.getLogger()
+
+
+class AWSSigV4RequestGenerator(AuthBase):
+    """
+    Signature Version 4 (SigV4) is the process to add authentication information to AWS API requests sent by HTTP.
+    For security, most requests to AWS must be signed with an access key. The access key consists of an access key
+    ID and secret access key, which are commonly referred to as your security credentials.
+
+    How Signature Version 4 works:
+
+    1. Create a canonical request.
+    2. Use the canonical request and additional metadata to create a string for signing.
+    3. Derive a signing key from your AWS secret access key. Then use the signing key, and the string from the previous
+       step, to create a signature.
+    4. Add the resulting signature to the HTTP request in a header
+    """
+
+
+    def __init__(self, **kwargs):
+        """
+        :param aws_service='es',
+        :param aws_access_key_id='YOUR_KEY_ID',
+        :param aws_secret_access_key='YOUR_SECRET',
+        :param aws_session_token='YOUR_SESSION_TOKEN'
+        :param aws_region='us-east-1',
+        :param aws_host='search-service.us-east-1.es.amazonaws.com',
+        """
+
+        self.aws_service = kwargs.get('aws_service')
+        self.aws_access_key_id = kwargs.get('aws_access_key_id')
+        self.aws_secret_access_key = kwargs.get('aws_secret_access_key')
+        self.aws_session_token = kwargs.get('aws_session_token')
+        self.aws_region = kwargs.get('aws_region')
+        self.aws_host = kwargs.get('aws_host')
+
+        if self.aws_service is None:
+            raise KeyError("Service is required")
+
+        if self.aws_access_key_id is None or self.aws_secret_access_key is None:
```

Comment:
```if any(True for x in [self.A, self.B] if x is None)```

Reply:
And I am not sure that in case one of them is missing you want to abandon the one that you have already pulled from kwargs and start fetching both from ``os.environ``.